### PR TITLE
DefaultTbLocalSubscriptionService WS single lock refactored by tenantId; TbEntityLocalSubsInfo performance optimizations for remove subscription

### PR DESCRIPTION
**Pull request by smatvienko-tb**
## DefaultTbLocalSubscriptionService WS single lock refactored by tenantId; TbEntityLocalSubsInfo performance optimizations for remove subscription

based on thread dump:

```
"DefaultWebSocketService-0-25" - Thread t@5330
   java.lang.Thread.State: WAITING
	at java.base@17.0.11/jdk.internal.misc.Unsafe.park(Native Method)
	- waiting to lock <7385cf25> (a java.util.concurrent.locks.ReentrantLock$NonfairSync) owned by "DefaultWebSocketService-0-45" t@104496
	at java.base@17.0.11/java.util.concurrent.locks.LockSupport.park(LockSupport.java:211)
	at java.base@17.0.11/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:715)
	at java.base@17.0.11/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:938)
	at java.base@17.0.11/java.util.concurrent.locks.ReentrantLock$Sync.lock(ReentrantLock.java:153)
	at java.base@17.0.11/java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:322)
	at org.thingsboard.server.service.subscription.DefaultTbLocalSubscriptionService.cancelAllSessionSubscriptions(DefaultTbLocalSubscriptionService.java:275)
	at org.thingsboard.server.service.ws.DefaultWebSocketService.cleanupSessionById(DefaultWebSocketService.java:827)
	at org.thingsboard.server.service.ws.DefaultWebSocketService.handleSessionEvent(DefaultWebSocketService.java:226)
	at jdk.internal.reflect.GeneratedMethodAccessor654.invoke(Unknown Source)
	at java.base@17.0.11/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@17.0.11/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:351)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
	at jdk.proxy2/jdk.proxy2.$Proxy340.handleSessionEvent(Unknown Source)
	at org.thingsboard.server.controller.plugin.TbWebSocketHandler.processInWebSocketService(TbWebSocketHandler.java:328)
	at org.thingsboard.server.controller.plugin.TbWebSocketHandler.afterConnectionClosed(TbWebSocketHandler.java:315)
	at org.springframework.web.socket.handler.WebSocketHandlerDecorator.afterConnectionClosed(WebSocketHandlerDecorator.java:85)
	at org.springframework.web.socket.handler.LoggingWebSocketHandlerDecorator.afterConnectionClosed(LoggingWebSocketHandlerDecorator.java:72)
	at org.springframework.web.socket.handler.ExceptionWebSocketHandlerDecorator.afterConnectionClosed(ExceptionWebSocketHandlerDecorator.java:78)
	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter.onClose(StandardWebSocketHandlerAdapter.java:144)
	at org.apache.tomcat.websocket.WsSession.fireEndpointOnClose(WsSession.java:714)
	at org.apache.tomcat.websocket.WsSession.doClose(WsSession.java:590)
	at org.apache.tomcat.websocket.WsSession.doClose(WsSession.java:556)
	at org.apache.tomcat.websocket.WsSession.close(WsSession.java:544)
	at org.springframework.web.socket.adapter.standard.StandardWebSocketSession.closeInternal(StandardWebSocketSession.java:235)
	at org.springframework.web.socket.adapter.AbstractWebSocketSession.close(AbstractWebSocketSession.java:144)
	at org.thingsboard.server.controller.plugin.TbWebSocketHandler.close(TbWebSocketHandler.java:554)
	at org.thingsboard.server.controller.plugin.TbWebSocketHandler$SessionMetaData.closeSession(TbWebSocketHandler.java:409)
	at org.thingsboard.server.controller.plugin.TbWebSocketHandler$SessionMetaData.sendPing(TbWebSocketHandler.java:397)
	at org.thingsboard.server.controller.plugin.TbWebSocketHandler.sendPing(TbWebSocketHandler.java:537)
	at org.thingsboard.server.service.ws.DefaultWebSocketService.lambda$sendPing$26(DefaultWebSocketService.java:923)
	at org.thingsboard.server.service.ws.DefaultWebSocketService$$Lambda$4375/0x00007f663dc94000.run(Unknown Source)
	at java.base@17.0.11/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1375)
	at java.base@17.0.11/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base@17.0.11/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base@17.0.11/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base@17.0.11/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base@17.0.11/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)

   Locked ownable synchronizers:
	- None


"ws-entity-sub-callback-58-thread-1" - Thread t@329
   java.lang.Thread.State: WAITING
	at java.base@17.0.11/jdk.internal.misc.Unsafe.park(Native Method)
	- waiting to lock <7385cf25> (a java.util.concurrent.locks.ReentrantLock$NonfairSync) owned by "DefaultWebSocketService-0-45" t@104496
	at java.base@17.0.11/java.util.concurrent.locks.LockSupport.park(LockSupport.java:211)
	at java.base@17.0.11/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:715)
	at java.base@17.0.11/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:938)
	at java.base@17.0.11/java.util.concurrent.locks.ReentrantLock$Sync.lock(ReentrantLock.java:153)
	at java.base@17.0.11/java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:322)
	at org.thingsboard.server.service.subscription.DefaultTbLocalSubscriptionService.addSubscription(DefaultTbLocalSubscriptionService.java:200)
	at jdk.internal.reflect.GeneratedMethodAccessor376.invoke(Unknown Source)
	at java.base@17.0.11/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@17.0.11/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:351)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
	at jdk.proxy2/jdk.proxy2.$Proxy413.addSubscription(Unknown Source)
	at org.thingsboard.server.service.subscription.TbAbstractDataSubCtx$$Lambda$4120/0x00007f663dbf9cc8.accept(Unknown Source)
	at java.base@17.0.11/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.thingsboard.server.service.subscription.TbAbstractDataSubCtx.createSubscriptions(TbAbstractDataSubCtx.java:158)
	at org.thingsboard.server.service.subscription.TbAbstractDataSubCtx.createLatestValuesSubscriptions(TbAbstractDataSubCtx.java:138)
	at org.thingsboard.server.service.subscription.DefaultTbEntityDataSubscriptionService$2.onSuccess(DefaultTbEntityDataSubscriptionService.java:691)
	at org.thingsboard.server.service.subscription.DefaultTbEntityDataSubscriptionService$2.onSuccess(DefaultTbEntityDataSubscriptionService.java:678)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1137)
	at java.base@17.0.11/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base@17.0.11/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base@17.0.11/java.lang.Thread.run(Thread.java:840)
```

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java [hunk 1]
```diff
@@ -266,18 +280,24 @@ public void cancelSubscription(String sessionId, int subscriptionId) {
...
-        subsLock.lock();
+        Lock subsLock = null;
         try {
             Map<Integer, TbSubscription<?>> sessionSubscriptions = subscriptionsBySessionId.remove(sessionId);
             if (sessionSubscriptions != null) {
                 for (TbSubscription<?> subscription : sessionSubscriptions.values()) {
+                    if (subsLock == null) {
+                    subsLock = getSubsLock(subscription.getTenantId());
+                    subsLock.lock();
+                    }
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java:291]: Please fix formatting

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java [hunk 2]
```diff
@@ -266,18 +280,24 @@ public void cancelSubscription(String sessionId, int subscriptionId) {
     public void cancelAllSessionSubscriptions(String sessionId) {
         log.debug("[{}] Going to remove session subscriptions.", sessionId);
         List<SubscriptionModificationResult> results = new ArrayList<>();
-        subsLock.lock();
+        Lock subsLock = null;
         try {
             Map<Integer, TbSubscription<?>> sessionSubscriptions = subscriptionsBySessionId.remove(sessionId);
             if (sessionSubscriptions != null) {
                 for (TbSubscription<?> subscription : sessionSubscriptions.values()) {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java:287]: Isn't it too late to grab the lock? As I understand, sessionSubscriptions might be concurrently modified at this point (by `cancelSubscription`)

Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java:287]: We need to pass TenantId as an argument to `cancelAllSessionSubscriptions` and other similar methods that do not have the TenantId

File: application/src/main/java/org/thingsboard/server/service/subscription/TbAlarmsSubscription.java [hunk 3]
```diff
@@ -36,12 +36,8 @@ public TbAlarmsSubscription(String serviceId, String sessionId, int subscription
     }
 
     @Override
-    public boolean equals(Object o) {
-        return super.equals(o);
+    protected boolean canEqual(final Object other) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbAlarmsSubscription.java:39]: I don't see how this is related to the scope of PR

File: application/src/main/java/org/thingsboard/server/service/subscription/TbEntityLocalSubsInfo.java [hunk 4]
```diff
@@ -137,8 +135,30 @@ public TbEntitySubEvent remove(TbSubscription<?> sub) {
             return toEvent(ComponentLifecycleEvent.DELETED);
         }
         TbSubscriptionsInfo oldState = state.copy();
-        TbSubscriptionsInfo newState = new TbSubscriptionsInfo();
+
+        //copy unchanged state only
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/TbEntityLocalSubsInfo.java:139]: Add comprehensive tests for these changes, or move them to different PR

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java [hunk 5]
```diff
@@ -176,6 +177,10 @@ public void onCoreStartupMsg(TransportProtos.CoreStartupMsg coreStartupMsg) {
         });
     }
 
+    Lock getSubsLock(TenantId tenantId) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java:180]: I suggest adding a class *EntityLocksStorage* that will use the *ConcurrentReferenceHashMap* and encapsulate a single operation with the locks—get(EntityId id). You will be able to cover it with tests (if needed). Then, the code will be a bit simpler, and we may reuse it in the future. Also, you will not need the get methods without the *private* modifier.

File: application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java [hunk 6]
```diff
@@ -219,14 +225,18 @@ public void onSubEventCallback(UUID entityId, int seqNumber, TbEntityUpdatesInfo
         log.debug("[{}][{}] Processing sub event callback: {}.", entityId, seqNumber, entityUpdatesInfo);
         entityUpdates.put(entityId, entityUpdatesInfo);
         Set<TbSubscription<?>> pendingSubs = null;
-        subsLock.lock();
+        Lock subsLock = null;
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/subscription/DefaultTbLocalSubscriptionService.java:228]: TbEntitySubEvent has TenantId
